### PR TITLE
Support Cassandra Storage in v3

### PR DESCRIPTION
**Pull request by mrproliu**
Reimplementing Cassandra Storage in Zipkin Using SkyWalking OAP Core.

## Data Tables
1. **Table Creation**: In the previous version, tables were created using predefined CQL files. Now, they would be dynamically created based on specific Java files (automatically checked and created when Zipkin startup).
2. **Table Names and Fields**: The tables have changed due to differences in implementation principles with the original Zipkin. The corresponding relationships are as follows:
    * span: [ZipkinSpanRecord.java](https://github.com/apache/skywalking/blob/master/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/zipkin/ZipkinSpanRecord.java)
    * trace_by_service_span: Utilizes the Span table.
    * trace_by_service_remote_service: Utilizes the Span table.
    * span_by_service: [ZipkinServiceSpanTraffic.java](https://github.com/apache/skywalking/blob/master/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/zipkin/ZipkinServiceSpanTraffic.java)
    * remote_service_by_service: Utilizes the span_by_service table.
    * autocomplete_tags: Utilizes the additional table in **Span**, as shown [here](https://github.com/apache/skywalking/blob/master/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/zipkin/ZipkinSpanRecord.java#L169).
3. **Field Types**: Previously, custom data types were used to store `endpoint` and `annotation` information. Now, these would be converted into `text` and stored across multiple fields.
4. **Table Storage**: Previously, all data were stored in their respective standalone tables. Now, tables are created on a daily basis for data insertion and querying.

## Configuration

1. **TTL (Time To Live)**: In the previous version of Zipkin, the `default_time_to_live` field was predefined in the tables to allow the database to delete data. In the new version, data would be periodically cleared from database tables in Zipkin according to the DATA_TTL configuration.
2. **Auto Complete Cardinality**: This configuration has been removed in the new version, and no longer has any restrictions.

### CODE REVIEW

**Comment by wu-sheng** [comment]
Let's put some summary about the feature here before the review.

File: zipkin-server/server-core/src/main/java/zipkin/server/core/CoreModuleProvider.java
Comment (wu-sheng) [inline zipkin-server/server-core/src/main/java/zipkin/server/core/CoreModuleProvider.java]: I think we still need cluster mode due to TTL task.

**Comment by anuraaga** [comment]
I haven't been following that well but skimmed the description and wanted to ask, if the tables change significantly as I think they did, it seems like a big migration is needed by users. If a migration would be needed, is there a reason to develop Zipkin v3 vs asking users to switch to skywalking oap? I reread JC's email and Bas's issue comment and don't think I saw that addressed.

**Comment by wu-sheng** [comment]
> I haven't been following that well but skimmed the description and wanted to ask, if the tables change significantly as I think they did, it seems like a big migration is needed by users. If a migration would be needed, is there a reason to develop Zipkin v3 vs asking users to switch to skywalking oap? I reread JC's email and Bas's issue comment and don't think I saw that addressed.

I think there was a discussion on the Zipkin admin mail list(I am not in that list).

About immigration, I think the general idea is not to keep the DB schema unchanged, we have new implementation in the OAP for MySQL/PostgreSQL/Elasticsearch/OpenSearch. And Cassandra is following the abstract.

**Comment by anuraaga** [comment]
> About immigration, I think the general idea is not to keep the DB schema unchanged,

Thanks - indeed if this is the general idea, I wonder if all of these schema changes will be accompanied by some auto migration logic. If so, it's probably better to start by writing the migration logic before the server logic since it's always possible to run into impossible migrations.

If not, I still wonder, what will Zipkin v3 offer over switching to Skywalker for Zipkin users? I think it could be helpful to have a bullet list for that.

File: zipkin-server/server-core/src/main/java/zipkin/server/core/CoreModuleProvider.java
Comment (mrproliu) [reply zipkin-server/server-core/src/main/java/zipkin/server/core/CoreModuleProvider.java]: Sure, I will create other PR to add the cluster mode.

**Comment by jcchavezs** [comment]
Although we aim to provide a drop in replacement for zipkin server we
discussed data retrocompatibility and decided to not to. Given the high
throughput and ingestion volume of such data, schema reconciliation can be
a pain.

What we agreed was that old data can't be consumed. If there are popular
demand on this (which I believe will be) maybe we can write tooling to
migrate data from one source to another so old data is also available in
new schema. Another option is that before one moves to the new server you
can have intermediate time where you leverage a OTel collector and write to
the two servers and progressively start moving to use new server.

**Comment by wu-sheng** [comment]
Personally, I would say this kind of immigration is not worth so much effect. Because, the TTL of spans is short, such as days. In some cases, the data would lost the value before the immigration is done.

The reason for different schema designs in SkyWalking, is from what we learned from large-scale deployment. As SkyWalking agent + OAP usually work for 100% sampling. We created more concepts to help query(including SkyWalking UI and Lens UI) to have better responses.
As JC said, this is not unchangeable. We haven't began to gather feedback as currently, we are only working on this branch rather than main branch, and finish the first preview version, which makes everything work.

> If not, I still wonder, what will Zipkin v3 offer over switching to Skywalker for Zipkin users? I think it could be helpful to have a bullet list for that.

This PR is the value. SkyWalking as an APM is impossible to work with Cassandra successfully, but Zipkin server can.
Another thing like GraalVM support. SkyWalking OAP would have a long way(one year, even longer) to support GraalVM. But here as Zipkin v3, this should be able to support soon. OAP has many analysis relative things, which is hard to immigrate to run on GraalVM.

One of the key point is Zipkin VS SkyWalking OAP is that, OAP is too complex for Zipkin users. It support all cases for an observability team, but Zipkin v3 could be much lighter, more stable and focus on tracing relative things. Meanwhile, we could share the experiences on both sides.

**Comment by anuraaga** [comment]
> write to the two servers and progressively start moving to use new server.

Yeah I assume that new server for new data, old server for old until deletion will be the practical approach vs trying any migration scripts.

So I'm still wondering why is this Zipkin? From what I follow in this thread, skywalking being more than tracing can actually impose limitations such as not working with Cassandra. That also seems weird to me though because if skywalking can't support Cassandra, there should be no need to change the Zipkin data model on Cassandra. If it's intended to be tracing-only Skywalking, isn't it better to still be part of Skywalking?

**Comment by wu-sheng** [comment]
> So I'm still wondering why is this Zipkin? From what I follow in this thread, skywalking being more than tracing can actually impose limitations such as not working with Cassandra. That also seems weird to me though because if skywalking can't support Cassandra, there should be no need to change the Zipkin data model on Cassandra. If it's intended to be tracing-only Skywalking, isn't it better to still be part of Skywalking?

Whether SkyWalking community likes this, it would take a long time to make another decision from a totally another perspective. Zipkin server has this storage option, let's keep it supported.

And, I think more discussion could be made in another place. 

Let's keep this PR context clear to the Cassandra itself.

**Comment by wu-sheng** [comment]
@mrproliu Please sync and avoid the conflicts as the bugs have been fixed through another PR.

**Comment by wu-sheng** [comment]
Let's wait for @michaelsembwever back from vacation and do the review about Cassandra codes and table structure. He knows Cassandra much better than I could ever be.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/EmptyDAO.java
Comment (wu-sheng) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/EmptyDAO.java]: Could you create a NotImplementedException? And throw from here?
It is better to see a clear exception rather than NPE or other strange behavior as there is nothing executed.
After all, these methods should not be called.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java:139]: neither SimpleStrategy nor rf=1 is appropriate for any production system. NetworkTopologyStrategy and  rf=3 is the minimum for production.

**Comment by michaelsembwever** [comment]
> In the previous version, tables were created using predefined CQL files.

Not having these makes it much harder to review the patch, and to benchmark it.
This needs the benchmark profiles so to be able to validate the scalability and superior performance characteristics of a Cassandra backend.  For example in master this is traces-stress.yaml and and trace_by_service_span-stress.yaml and span_by_service-stress.yaml files.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java]: why are we using jdbc.SQLBuilder here? the java-driver used natively will be safer and cleaner.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/BatchCQLExecutor.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/BatchCQLExecutor.java]: don't do batch statements.  and don't store then flush writes.  this is misunderstanding C*'s design and strengths.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java]: if we are bucketing by table do we still need TimeWindowCompactionStrategy? 

what is the performance impact of daily creating and dropping tables ? 

The correct (best performance and availability) design here is to define the time bucket inside the one table in its primary key, define the TTL at the table/write level, and let TimeWindowCompactionStrategy do what it's best at.  Here it feels like we are re-inventing the wheel at the application layer.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: don't use `IN ` against. the primary key. async concurrent requests instead.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java]: 1. Avoid indexes, especially legacy 2i.  (v2 uses SASI, if you must better is to wait for 5.0 and use SAI, at minimum create a design that can still run without the index – true scalability as well as availability when an index needs to be rebuilt)
2. `indexExists(…)` does not work in a distributed environment.  it can be used as an optimisation but cannot be relied upon. use `IF NOT EXISTS`.  use the optimisation everywhere though (CREATE TABLE too).

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraCqlExecutor.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraCqlExecutor.java]: use the java-driver natively.  and don't use `IN` against the primary key.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java]: is the primary key specified in the where clause here?
allow filtering without the primary key specified is a full-table scan and will not work.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java
Comment (michaelsembwever) [inline zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java]: is the primary key specified in the where clause here?
allow filtering without the primary key specified is a full-table scan and will not work.

**Review by michaelsembwever — CHANGES_REQUESTED** [review]

**Comment by michaelsembwever** [comment]
To summarise
 - Cassandra storage is for deployments that scale.  It can and should be designed to
  -- be as performant as, if not more than, other storage choices at no/low scale (bc C* is the preferred/only available db)
  --  work where the other storage won't
  -- scale linearly, so operators know it will continue to work as load increases, and costs over scale are predictable
  -- to validate any of this we need the benchmark profiles
- batch statements, and storing and flushing of writes, should not be done
- time bucketing at the primary key, and TTL in the table/write, is superior (and what TWCS is designed for)
- allow filtering used without specifying the primary key must be avoided
- using the IN clause against the primary key must be avoided
- legacy secondary indexes must be avoided
- SASI indexes can be optionally used, but the application needs to work (degraded mode) without them
- C* 5.0 offers SAI indexes, superior to SASI, ok to use (but still try to design without them)
- keyspace replication config must be configuration (and not be Simple nor rf<3 in any production)
- you cannot rely on read+write coordination at the application layer: e.g. `indexExists(…)`


If you have integration tests, you can also rely upon C* server-side guardrails to check you're designing and using correctly.  (Such a guardrails definition would also be valuable advice to production operators.)

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java]: Sure, the `ALLOW FILTERING` has been deleted.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraTagAutocompleteDAO.java]: Sure, the `ALLOW FILTERING` has been deleted.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraCqlExecutor.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraCqlExecutor.java]: Understood. I have divided each CQL query for execution.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java]: > Avoid indexes, especially legacy 2i. (v2 uses SASI, if you must better is to wait for 5.0 and use SAI, at minimum create a design that can still run without the index – true scalability as well as availability when an index needs to be rebuilt)

Updated, still using the SASI index for now. I could update the SAI index when Cassandra v5.0 is ready. 

> indexExists(…) does not work in a distributed environment. it can be used as an optimisation but cannot be relied upon. use IF NOT EXISTS. use the optimisation everywhere though (CREATE TABLE too).

Update to using the pre-defined table CQL file.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: Updated to using multiple equals query.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: I have updated to use of the latest secondary table in the v2 for searching.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: I have updated to use of the latest secondary table in the v2 for searching.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: I have updated to use of the latest secondary table in the v2 for searching.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java]: I have updated to use of the latest secondary table in the v2 for searching.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java]: Sure, I have updated to use the original Cassandra table settings, nothing more.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/BatchCQLExecutor.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/BatchCQLExecutor.java]: Understood. I have divided each CQL query for execution.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java]: Deleted.

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java:139]: These are the default settings in the [Zipkin CQL file](https://github.com/openzipkin/zipkin/blob/master/zipkin-storage/cassandra/src/main/resources/zipkin2-schema.cql#L1-L3), should I updated?

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java
Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraHistoryDeleteDAO.java]: Sure, I have updated to use the original Cassandra table settings, nothing more.

**Comment by wu-sheng** [comment]
@michaelsembwever Could you review for another round?

File: zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java
Comment (michaelsembwever) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java:139]: they are appropriate for first-time / dev usage. but not appropriate for production usage. 

it needs to be configurable.
and ideally the application should log a warning when rf<3 saying it's not appropriate for production.

Comment (mrproliu) [reply zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraClient.java:139]: Yes, if rf<3 would be got a warning log. What do you mean configurable?

**Review by wu-sheng — APPROVED** [review]
@michaelsembwever I am going to merge this first, to unblock the Cassandra relative changes at the dependency repo side.

When you have time, feel free to review, and we will follow your recommendations.

**Pull request merged**
